state/migration_import.go

```diff
@@ -1463,17 +1451,38 @@ func (i *importer) makeCharmOrigin(a description.Application) (*CharmOrigin, err
 		Channel:      p.Channel,
 	}
 
+	// NOTE(jack-w-shaw) Due to buggy clients and servers we cannot trust the
+	// charm origin is accurate. So when we import a charm origin, attempt to
+	// construct as much as possible from the charm url
```

Comment:
This is problematic as we're trying to reduce dependence on the charmurl. The comment is also misleading because it's only partially true, we must trust the id/hash/channel/platform in reallity. Let's fix the things we know about.

Reply:
There isn't really much we can do about this usage of charm url at this stage. We can't fully remove dependence on charm url until we have other ways of getting the data stored, meaning that during migrations/upgrades we will need to parse this data from existing charm urls to fill it in

You're right about id/hash though. I'll add a bit more specificity to the comment. The problem here isn't inaccurate data, but potentially out of date data. Which should be reflected in my comment

Reply:
Well, I guess it could potentially be inaccurate. There is nothing stopping someone from using a hacked client to send incorrect data into the server, there's nothing in the server to stop this. It's more that this is unlikely and we don't have any other option